src/ssl_cert_paths_discover.c

```diff
@@ -0,0 +1,152 @@
+ *
+ * Redistribution and use in source and binary forms, with or
+ * without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above
+ *    copyright notice, this list of conditions and the
+ *    following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above
+ *    copyright notice, this list of conditions and the following
+ *    disclaimer in the documentation and/or other materials
+ *    provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY <COPYRIGHT HOLDER> ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * <COPYRIGHT HOLDER> OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
+ * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <dirent.h>
+#include <sys/stat.h>
+#include <openssl/ssl.h>
+#include "ssl_cert_paths_discover.h"
+#include "tt_static.h"
+
+static int
+is_dir_empty(const char *dir_path)
+{
+	DIR *dir = opendir(dir_path);
+	if (!dir)
+		return 1;
+
+	struct dirent *ent;
+	int is_empty = 1;
+	while ((ent = readdir(dir))) {
+		if (strcmp(ent->d_name, ".") && strcmp(ent->d_name, "..")) {
+			is_empty = 0;
+			break;
+		}
+	}
+
+	closedir(dir);
+	return is_empty;
+}
+
+const char *default_cert_dir_paths[] = {
+	/* Fedora/RHEL/Centos */
+	"/etc/pki/tls/certs",
+	/* Debian/Ubuntu/Gentoo etc. (OpenSuse) */
+	"/etc/ssl/certs",
+	/* FreeBSD */
+	"/usr/local/share/certs",
+	/* NetBSD */
+	"/etc/openssl/certs",
+	/* macOS */
+	"/private/etc/ssl/certs",
+	"/usr/local/etc/openssl@1.1/certs",
+	"/usr/local/etc/openssl@1.0/certs",
+	"/usr/local/etc/openssl/certs",
+	NULL
+};
+
+const char *default_cert_file_paths[] = {
+	/* Fedora/RHEL 6 */
+	"/etc/pki/tls/certs/ca-bundle.crt",
+	/* CentOS/RHEL 7/8 */
+	"/etc/pki/ca-trust/extracted/pem/tls-ca-bundle.pem",
+	/* Debian/Ubuntu/Gentoo etc. */
+	"/etc/ssl/certs/ca-certificates.crt",
+	/* OpenSUSE */
+	"/etc/ssl/ca-bundle.pem",
+	/* FreeBSD */
+	"/usr/local/share/certs/ca-root-nss.crt",
+	/* macOS */
+	"/private/etc/ssl/cert.pem",
+	"/usr/local/etc/openssl@1.1/cert.pem",
+	"/usr/local/etc/openssl@1.0/cert.pem",
+	NULL
+};
+
+int
+ssl_cert_paths_discover(int overwrite)
+{
+	int rc = -1;
+	struct stat sb;
+	size_t dir_buf_end = 0;
+	size_t initial_size = 32;
+	size_t buf_size = initial_size * sizeof(char);
```

Comment:
sizeof() returns size in chars (not in bytes). There is no point in "* sizeof(char)".

Reply:
Fixed.
I've used sizeof(char) for clarity, but you are right it can be omitted